# Spam Detector

In [1]:
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.metrics import accuracy_score

## Retrieve the Data

The data is located at [https://static.bc-edx.com/ai/ail-v-1-0/m13/challenge/spam-data.csv](https://static.bc-edx.com/ai/ail-v-1-0/m13/challenge/spam-data.csv)

Dataset Source: [UCI Machine Learning Library](https://archive.ics.uci.edu/dataset/94/spambase)

Import the data using Pandas. Display the resulting DataFrame to confirm the import was successful.

In [2]:
# Import the data
data = pd.read_csv("https://static.bc-edx.com/ai/ail-v-1-0/m13/challenge/spam-data.csv")
data.head()

word_freq_make  word_freq_address  word_freq_all  word_freq_3d  \
0            0.00               0.64           0.64           0.0   
1            0.21               0.28           0.50           0.0   
2            0.06               0.00           0.71           0.0   
3            0.00               0.00           0.00           0.0   
4            0.00               0.00           0.00           0.0   

   word_freq_our  word_freq_over  word_freq_remove  word_freq_internet  \
0           0.32            0.00              0.00                0.00   
1           0.14            0.28              0.21                0.07   
2           1.23            0.19              0.19                0.12   
3           0.63            0.00              0.31                0.63   
4           0.63            0.00              0.31                0.63   

   word_freq_order  word_freq_mail  ...  char_freq_;  char_freq_(  \
0             0.00            0.00  ...         0.00        0.000   
1             0.00            0.94  ...         0.00        0.132   
2             0.64            0.25  ...         0.01        0.143   
3             0.31            0.63  ...         0.00        0.137   
4             0.31            0.63  ...         0.00        0.135   

   char_freq_[  char_freq_!  char_freq_$  char_freq_#  \
0          0.0        0.778        0.000        0.000   
1          0.0        0.372        0.180        0.048   
2          0.0        0.276        0.184        0.010   
3          0.0        0.137        0.000        0.000   
4          0.0        0.135        0.000        0.000   

   capital_run_length_average  capital_run_length_longest  \
0                       3.756                          61   
1                       5.114                         101   
2                       9.821                         485   
3                       3.537                          40   
4                       3.537                          40   

   capital_run_length_total  spam  
0                       278     1  
1                      1028     1  
2                      2259     1  
3                       191     1  
4                       191     1  

[5 rows x 58 columns]

## Predict Model Performance

You will be creating and comparing two models on this data: a Logistic Regression, and a Random Forests Classifier. Before you create, fit, and score the models, make a prediction as to which model you think will perform better. You do not need to be correct! 

Write down your prediction in the designated cells in your Jupyter Notebook, and provide justification for your educated guess.

*Replace the text in this markdown cell with your predictions, and be sure to provide justification for your guess.*

This is a random guess that the logistic regression model will be better. That is my prediction. Educated guess will be hard at this point as I am not sure what to say.

## Split the Data into Training and Testing Sets

In [13]:
# Create the labels set `y` and features DataFrame `X`
# Create the labels set `y`
y = data['spam']

# Create the features DataFrame `X`
X = data.drop(columns=['spam'])

# Display the first few rows of X and y to confirm
#print(X.head())
#print(y.head())

In [4]:
# Check the balance of the labels variable (`y`) by using the `value_counts` function.
# Check the balance of the labels
label_balance = y.value_counts()

# Display the balance of the labels
print(label_balance)

spam
0    2788
1    1813
Name: count, dtype: int64


In [5]:
from sklearn.model_selection import train_test_split

# Split the data
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

# Display the shapes of the resulting datasets to confirm the split
print(f'X_train shape: {X_train.shape}')
print(f'X_test shape: {X_test.shape}')
print(f'y_train shape: {y_train.shape}')
print(f'y_test shape: {y_test.shape}')
X_train.head()

X_train shape: (3680, 57)
X_test shape: (921, 57)
y_train shape: (3680,)
y_test shape: (921,)


word_freq_make  word_freq_address  word_freq_all  word_freq_3d  \
1370            0.09                0.0           0.09           0.0   
3038            0.00                0.0           0.00           0.0   
2361            0.00                0.0           2.43           0.0   
156             0.00                0.0           0.00           0.0   
2526            0.00                0.0           0.00           0.0   

      word_freq_our  word_freq_over  word_freq_remove  word_freq_internet  \
1370           0.39            0.09              0.09                0.00   
3038           0.00            0.00              0.00                0.00   
2361           0.00            0.00              0.00                0.00   
156            1.31            0.00              1.31                1.31   
2526           0.00            0.00              0.00                0.00   

      word_freq_order  word_freq_mail  ...  word_freq_conference  char_freq_;  \
1370             0.19            0.29  ...                   0.0        0.000   
3038             0.00            0.00  ...                   0.0        0.124   
2361             0.27            0.00  ...                   0.0        0.000   
156              1.31            1.31  ...                   0.0        0.000   
2526             0.00            0.00  ...                   0.0        0.000   

      char_freq_(  char_freq_[  char_freq_!  char_freq_$  char_freq_#  \
1370        0.139          0.0        0.310        0.155          0.0   
3038        0.124          0.0        0.000        0.000          0.0   
2361        0.344          0.0        0.000        0.000          0.0   
156         0.000          0.0        0.117        0.117          0.0   
2526        0.000          0.0        0.000        0.000          0.0   

      capital_run_length_average  capital_run_length_longest  \
1370                       6.813                         494   
3038                       1.800                           8   
2361                       2.319                          12   
156                       48.500                         186   
2526                       2.307                           8   

      capital_run_length_total  
1370                      1458  
3038                        45  
2361                       167  
156                        291  
2526                        30  

[5 rows x 57 columns]

## Scale the Features

Use the `StandardScaler` to scale the features data. Remember that only `X_train` and `X_test` DataFrames should be scaled.

In [6]:
from sklearn.preprocessing import StandardScaler

# Create the StandardScaler instance

scaler = StandardScaler()

In [7]:
# Fit the Standard Scaler with the training data
# Fit the scaler to the training data

# Fit the scaler to X_train and transform it
X_scaled = scaler.fit(X_train)

# After fitting, you can transform the training and test data if needed
X_train_scaled = X_scaled.transform(X_train)
X_test_scaled = X_scaled.transform(X_test)

## Create and Fit a Logistic Regression Model

Create a Logistic Regression model, fit it to the training data, make predictions with the testing data, and print the model's accuracy score. You may choose any starting settings you like. 

In [17]:
# Train a Logistic Regression model and print the model score
from sklearn.linear_model import LogisticRegression

from sklearn.metrics import accuracy_score

# Create a Logistic Regression model
model = LogisticRegression(random_state=1)
print(model)

# Train the model using the scaled training data
model.fit(X_train_scaled, y_train)

# Predict the labels for the test data
y_pred = model.predict(X_test_scaled)

# Print the accuracy score of the model
accuracy = accuracy_score(y_test, y_pred)
print(f'Model Accuracy: {accuracy:.4f}')
print(y_pred)





LogisticRegression(random_state=1)
Model Accuracy: 0.9197
[0 0 0 1 0 1 0 0 0 1 0 0 0 1 0 0 0 0 1 0 0 1 0 0 1 0 1 0 1 1 1 1 0 1 0 0 1
 1 0 1 1 0 0 0 0 1 1 0 0 1 1 0 0 0 0 1 0 0 0 0 1 0 1 1 1 0 0 0 0 0 0 1 0 1
 1 0 0 0 0 1 0 0 0 0 0 0 1 1 1 0 1 1 0 0 0 0 1 0 1 1 0 1 0 1 1 0 0 1 1 1 1
 1 1 0 1 1 0 0 0 0 1 0 1 0 1 0 0 0 1 0 0 1 1 0 0 0 1 0 1 1 1 1 0 0 1 1 0 0
 0 0 0 1 1 1 0 0 0 1 1 0 1 1 1 0 1 0 1 0 0 1 0 0 0 0 1 0 1 0 0 1 0 0 0 0 1
 1 0 0 0 0 1 0 1 0 0 0 0 0 0 0 0 0 0 0 0 1 0 0 0 1 0 1 0 0 0 0 0 0 0 0 0 1
 1 1 1 0 1 0 0 0 0 0 0 0 0 0 0 0 1 1 0 1 1 1 0 0 0 1 0 0 0 1 0 0 0 0 0 1 1
 0 1 0 1 0 1 1 1 0 0 0 1 0 0 0 0 0 1 0 0 0 0 1 0 0 1 0 0 1 0 1 0 1 0 0 1 1
 0 0 0 0 0 1 0 0 0 0 1 1 1 0 1 0 1 1 0 0 0 1 1 0 0 0 0 0 0 1 1 1 1 0 0 0 1
 0 0 1 0 1 0 0 0 0 0 0 1 0 0 0 1 1 0 1 0 0 0 1 0 1 1 1 0 1 1 0 0 0 0 0 0 0
 1 1 0 1 0 0 0 0 1 0 1 1 0 0 1 0 1 0 0 0 0 0 1 1 0 1 0 1 0 0 1 0 0 0 1 1 0
 1 1 1 0 0 1 0 0 1 0 0 0 1 0 1 0 0 0 0 1 1 0 0 0 1 0 0 1 1 0 0 1 0 0 0 1 1
 0 1 0 0 0 1 0 0 0 1 0 0 1 1 1 1 1 1 0 0 1

In [20]:
# Make and save testing predictions with the saved logistic regression model using the test data

# Review the predictions

from sklearn.ensemble import RandomForestClassifier
import joblib

# Create and train the Logistic Regression model
model = RandomForestClassifier(random_state=1)
model.fit(X_train_scaled, y_train)
print(model)

y_pred = model.predict(X_test_scaled)








RandomForestClassifier(random_state=1)


In [21]:
accuracy_score(y_test,y_pred)

0.9565689467969598

## Create and Fit a Random Forest Classifier Model

Create a Random Forest Classifier model, fit it to the training data, make predictions with the testing data, and print the model's accuracy score. You may choose any starting settings you like. 

In [22]:
# Train a Random Forest Classifier model and print the model score?????????????
from sklearn.ensemble import RandomForestClassifier

from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import accuracy_score

# Step 1: Create the Random Forest Classifier instance
rf_model = RandomForestClassifier(random_state=42)
print (rf_model)


# Step 2: Train the model using the scaled training data
rf_model.fit(X_train_scaled, y_train)

# Step 3: Make predictions on the test data
y_pred_rf = rf_model.predict(X_test_scaled)

# Step 4: Calculate and print the accuracy score
accuracy_rf = accuracy_score(y_test, y_pred_rf)
print(f'Random Forest Classifier Accuracy: {accuracy_rf:.4f}')

RandomForestClassifier(random_state=42)
Random Forest Classifier Accuracy: 0.9555


## Evaluate the Models

Which model performed better? How does that compare to your prediction? Write down your results and thoughts in the following markdown cell.

*Replace the text in this markdown cell with your answers to these questions.*

The model that performed better was the Random Forest Classifier with the result of 0.9555 compared to the Logitic Regression model with the accuracy of 0.9197. At the begining of the assignment I had initially predicted the Logistic Regression.